### ISSUE TITLE
multi_scan occasionally crashes while trying to rename a folder after retrying a failed test

### ISSUE THREAD

**Issue by dp-1a** [lyndsey-ferguson/fastlane-plugin-test_center#169]
### New Issue Checklist

- [x] Updated `fastlane-plugin-test_center` to the latest version
- [x] I read the [README.md](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/blob/master/README.md)
- [x] I reviewed the [example(s)](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/blob/master/README.md) for the action(s) I am using
- [x] I have removed any sensitive data such as passwords, authentication tokens, or anything else I do not want to world to see

### Issue Description

`multi_scan` occasionally chrashes while trying to rename a folder after retrying a failed test. The function in which the error occurs seems to be `RetryingScanHelper#move_test_result_bundle_for_next_run` where it fails to rename the folder on line 306.

Currently we cannot reliably reproduce the issue therefore it seems to be some kind of edge case. 

##### Complete output when running fastlane, including the stack trace and command used

<details>
  <pre>Errno::ENOTEMPTY: [!] Directory not empty @ rb_file_s_rename - (/Users/runner/runners/2.160.0/work/1/s/fastlane/output/General_Tests-batch-1/General.test_result, /Users/runner/runners/2.160.0/work/1/s/fastlane/output/General_Tests-batch-1/General-2.test_result)
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-plugin-test_center-3.8.13/lib/fastlane/plugin/test_center/helper/multi_scan_manager/retrying_scan_helper.rb:306:in `rename'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-plugin-test_center-3.8.13/lib/fastlane/plugin/test_center/helper/multi_scan_manager/retrying_scan_helper.rb:306:in `move_test_result_bundle_for_next_run'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-plugin-test_center-3.8.13/lib/fastlane/plugin/test_center/helper/multi_scan_manager/retrying_scan_helper.rb:151:in `handle_test_failure'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-plugin-test_center-3.8.13/lib/fastlane/plugin/test_center/helper/multi_scan_manager/retrying_scan_helper.rb:113:in `after_testrun'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-plugin-test_center-3.8.13/lib/fastlane/plugin/test_center/helper/multi_scan_manager/retrying_scan.rb:63:in `rescue in run'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-plugin-test_center-3.8.13/lib/fastlane/plugin/test_center/helper/multi_scan_manager/retrying_scan.rb:51:in `run'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-plugin-test_center-3.8.13/lib/fastlane/plugin/test_center/helper/multi_scan_manager/retrying_scan.rb:45:in `run'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-plugin-test_center-3.8.13/lib/fastlane/plugin/test_center/helper/multi_scan_manager/test_batch_worker.rb:18:in `run'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-plugin-test_center-3.8.13/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:143:in `block in run_test_batches'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-plugin-test_center-3.8.13/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:140:in `each'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-plugin-test_center-3.8.13/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:140:in `each_with_index'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-plugin-test_center-3.8.13/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:140:in `run_test_batches'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-plugin-test_center-3.8.13/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:53:in `run'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-plugin-test_center-3.8.13/lib/fastlane/plugin/test_center/actions/multi_scan.rb:24:in `run'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-2.135.2/fastlane/lib/fastlane/runner.rb:256:in `block (2 levels) in execute_action'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-2.135.2/fastlane/lib/fastlane/actions/actions_helper.rb:50:in `execute_action'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-2.135.2/fastlane/lib/fastlane/runner.rb:235:in `block in execute_action'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-2.135.2/fastlane/lib/fastlane/runner.rb:227:in `chdir'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-2.135.2/fastlane/lib/fastlane/runner.rb:227:in `execute_action'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-2.135.2/fastlane/lib/fastlane/runner.rb:157:in `trigger_action_by_name'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-2.135.2/fastlane/lib/fastlane/fast_file.rb:159:in `method_missing'
  Fastfile:369:in `block (2 levels) in parsing_binding'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-2.135.2/fastlane/lib/fastlane/lane.rb:33:in `call'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-2.135.2/fastlane/lib/fastlane/runner.rb:49:in `block in execute'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-2.135.2/fastlane/lib/fastlane/runner.rb:45:in `chdir'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-2.135.2/fastlane/lib/fastlane/runner.rb:45:in `execute'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-2.135.2/fastlane/lib/fastlane/lane_manager.rb:56:in `cruise_lane'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-2.135.2/fastlane/lib/fastlane/command_line_handler.rb:36:in `handle'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-2.135.2/fastlane/lib/fastlane/commands_generator.rb:108:in `block (2 levels) in run'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-2.135.2/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:76:in `run!'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-2.135.2/fastlane/lib/fastlane/commands_generator.rb:349:in `run'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-2.135.2/fastlane/lib/fastlane/commands_generator.rb:41:in `start'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-2.135.2/fastlane/lib/fastlane/cli_tools_distributor.rb:119:in `take_off'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/gems/fastlane-2.135.2/bin/fastlane:23:in `<top (required)>'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/bin/fastlane:23:in `load'
  /Users/runner/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/bin/fastlane:23:in `<top (required)>'</pre>
</details>

### Environment

Running on macOS agent on Azure DevOps

**Comment by lyndsey-ferguson** [comment]
@dp-1a thank you for the Issue report. Can you include the output of `fastlane env` here?

**Comment by dp-1a** [comment]
Hi @lyndsey-ferguson,

attached the relevant parts of the fastlane file. The scenario seems only to happen when tests are being retried after a failure. Questions is who creates the initial folder during the tests run and why might it be already existing?

We are currently trying to split the execution of the UITests over multiple azure agents to speed up the execution.


### Stack

| Key                         | Value                                                            |
| --------------------------- | ---------------------------------------------------------------- |
| OS                          | 10.14.6                                                          |
| Ruby                        | 2.6.3                                                            |
| Bundler?                    | true                                                             |
| Git                         | git version 2.24.0                                               |
| Installation Source         | ~/runners/2.160.0/work/1/s/vendor/bundle/ruby/2.6.0/bin/fastlane |
| Host                        | Mac OS X 10.14.6 (18G1012)                                       |
| Ruby Lib Dir                | ~/hostedtoolcache/Ruby/2.6.3/x64/lib                             |
| OpenSSL Version             | OpenSSL 1.0.2t  10 Sep 2019                                      |
| Is contained                | false                                                            |
| Is homebrew                 | false                                                            |
| Is installed via Fabric.app | false                                                            |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/                      |
| Xcode Version               | 11.1                                                             |


### System Locale

| Variable | Value       |   
| -------- | ----------- | 
| LANG     | en_US.UTF-8 | 
| LC_ALL   | en_US.UTF-8 | 
| LANGUAGE |             |   


### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
ruby_version "2.3.0"
fastlane_version "2.130.0"
default_platform(:ios)

XCODE_VERSION = "~>10.0"

  desc "Tests a certain fraction of the whole test suite out of the derived data"
  desc "#### Example:"
  desc "```\nbundle exec fastlane fractionTest customer:XX --env general\n```"
  desc "#### Options:"
  desc " * **`customer`** (required): The customer label to use."
  desc " * **`test_type`**: Determines if either 'ui' or 'unit' tests should be run. Defaults to 'ui'"
  desc " * **`job_number`**: The current job number to determine which fraction of all tests should be run. Defaults to '1'"
  desc " * **`parallel_jobs`**: The number on how many parallel jobs the tests are distributed. Defaults to '1'"
  desc " * **`output_postfix`**: Postfixes the output directory. Defaults to ''"
  desc "#### Environments:"
  desc " * **`general`**: App identifier is `XXX`"
  lane :fractionTest do |options|
    # get all test classes and method names
    UI.header("Step: collect_tests")
    config = {
      :xctestrun => derived_testrun_path(DERIVED_DATA_PATH, ENV["MNCO_SCHEME"]),
      :invocation_based_tests => false
    }
    tests = ::Fastlane::Actions::TestsFromXctestrunAction.run(config)
    unitTests = tests["#{ENV["MNCO_SCHEME"]}_Tests"]
    uiTests = tests["#{ENV["MNCO_SCHEME"]}_UITests"]
    UI.message("Found '#{unitTests.count}' unit tests")
    UI.message("Found '#{uiTests.count}' UI tests")

    # determines which fraction of the tests should be tested
    UI.header("Step: determine_test_fraction")
    jobNumber = 1
    parallelJobs = 1
    if options[:job_number] && options[:parallel_jobs]
      jobNumber = options[:job_number].to_i
      parallelJobs = options[:parallel_jobs].to_i
    end
    UI.message("Running job '#{jobNumber}' out of '#{parallelJobs}' parallel jobs")

    # get output postfix
    outputPostfix = ""
    if options[:output_postfix]
      outputPostfix = "/#{options[:output_postfix]}"
    end

    # determine test type
    if options[:test_type].downcase == "unit"
      testing = unitTests
    else
      testing = uiTests
    end

    # calculating the size of each slice
    sliceSize = (testing.count/parallelJobs.to_f).ceil

    # run tests with multi scan which reruns certain test cases
    result = multi_scan(
      workspace: WORKSPACE,
      scheme: ENV["MNCO_SCHEME"],
      devices: TEST_DEVICES,
      clean: true,
      code_coverage: true,
      output_directory: "#{OUTPUT_DIRECTORY}#{outputPostfix}",
      derived_data_path: DERIVED_DATA_PATH,
      test_without_building: true,
      only_testing: testing.each_slice(sliceSize).to_a[jobNumber-1],
      try_count: 5,
      fail_build: false,
    )

    unless result[:failed_testcount].zero?
      UI.user_error!("There are #{result[:failed_testcount]} legitimate failing tests")
    end
  end
end

```
</details>

**No Appfile found**


### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.135.2 | âœ… Up-To-Date  |


### Loaded fastlane plugins:

| Plugin                      | Version | Update-Status      |
| --------------------------- | ------- | ------------------ |
| fastlane-plugin-badge       | 1.3.0   | âœ… Up-To-Date       |
| fastlane-plugin-appcenter   | 1.6.0   | ðŸš« Update available |
| fastlane-plugin-test_center | 3.8.13  | âœ… Up-To-Date       |


<details><summary><b>Loaded gems</b></summary>

| Gem                         | Version      |
| --------------------------- | ------------ |
| did_you_mean                | 1.3.0        |
| bundler                     | 2.0.2        |
| CFPropertyList              | 3.0.1        |
| concurrent-ruby             | 1.1.5        |
| i18n                        | 0.9.5        |
| minitest                    | 5.11.3       |
| thread_safe                 | 0.3.6        |
| tzinfo                      | 1.2.5        |
| activesupport               | 4.2.11.1     |
| public_suffix               | 2.0.5        |
| addressable                 | 2.7.0        |
| atomos                      | 0.1.3        |
| babosa                      | 1.0.3        |
| curb                        | 0.9.10       |
| fastimage                   | 2.1.7        |
| colored                     | 1.2          |
| highline                    | 1.7.10       |
| commander-fastlane          | 4.4.6        |
| dotenv                      | 2.7.5        |
| emoji_regex                 | 1.0.1        |
| excon                       | 0.68.0       |
| multipart-post              | 2.0.0        |
| faraday                     | 0.17.0       |
| unf_ext                     | 0.0.7.6      |
| unf                         | 0.1.4        |
| domain_name                 | 0.5.20190701 |
| http-cookie                 | 1.0.3        |
| faraday-cookie_jar          | 0.0.6        |
| faraday_middleware          | 0.13.1       |
| gh_inspector                | 1.1.3        |
| jwt                         | 2.1.0        |
| memoist                     | 0.16.1       |
| multi_json                  | 1.14.1       |
| os                          | 1.0.1        |
| signet                      | 0.11.0       |
| googleauth                  | 0.6.7        |
| httpclient                  | 2.8.3        |
| mime-types-data             | 3.2019.1009  |
| mime-types                  | 3.3          |
| declarative                 | 0.0.10       |
| declarative-option          | 0.1.0        |
| uber                        | 0.1.0        |
| representable               | 3.0.4        |
| retriable                   | 3.1.2        |
| google-api-client           | 0.23.9       |
| digest-crc                  | 0.4.1        |
| google-cloud-env            | 1.2.1        |
| google-cloud-core           | 1.3.2        |
| google-cloud-storage        | 1.16.0       |
| json                        | 2.2.0        |
| mini_magick                 | 4.9.5        |
| multi_xml                   | 0.6.0        |
| plist                       | 3.5.0        |
| rubyzip                     | 1.3.0        |
| security                    | 0.1.3        |
| naturally                   | 2.2.0        |
| simctl                      | 1.6.6        |
| slack-notifier              | 2.3.2        |
| terminal-notifier           | 2.0.0        |
| unicode-display_width       | 1.6.0        |
| terminal-table              | 1.8.0        |
| tty-screen                  | 0.7.0        |
| tty-cursor                  | 0.7.0        |
| tty-spinner                 | 0.9.1        |
| word_wrap                   | 1.0.0        |
| claide                      | 1.0.3        |
| colored2                    | 3.1.2        |
| nanaimo                     | 0.2.6        |
| xcodeproj                   | 1.13.0       |
| rouge                       | 2.0.7        |
| xcpretty                    | 0.3.0        |
| xcpretty-travis-formatter   | 1.0.0        |
| badge                       | 0.11.0       |
| fuzzy_match                 | 2.0.4        |
| nap                         | 1.1.0        |
| cocoapods-core              | 1.7.5        |
| cocoapods-deintegrate       | 1.0.4        |
| cocoapods-downloader        | 1.2.2        |
| cocoapods-plugins           | 1.0.0        |
| cocoapods-search            | 1.0.0        |
| cocoapods-stats             | 1.1.0        |
| netrc                       | 0.11.0       |
| cocoapods-trunk             | 1.3.1        |
| cocoapods-try               | 1.1.0        |
| escape                      | 0.0.4        |
| fourflusher                 | 2.3.1        |
| molinillo                   | 0.6.6        |
| ruby-macho                  | 1.4.0        |
| cocoapods                   | 1.7.5        |
| colorize                    | 0.8.1        |
| fastlane-plugin-appcenter   | 1.6.0        |
| fastlane-plugin-badge       | 1.3.0        |
| xctest_list                 | 1.1.8        |
| fastlane-plugin-test_center | 3.8.13       |
| xcode-install               | 2.6.2        |
</details>


*generated on:* **2019-11-15**
</details>

**Comment by lyndsey-ferguson** [comment]
@dp-1a so, that folder `/Users/runner/runners/2.160.0/work/1/s/fastlane/output/General_Tests-batch-1/General-2.test_result` is created by `multi_scan` in preparation for the next test run. It is doing that because `xcodebuild` via `scan` will create `/Users/runner/runners/2.160.0/work/1/s/fastlane/output/General_Tests-batch-1/General.test_result` and we need a new one for each test-run.

I wanted to see your `Fastfile` to see what options you were providing to see if there is anything suspicious: nothing.

Something else is creating that `General-2.test_result` folder. Normally, I would suspect that a parallel process is creating this folder. Given the `parallelJobs` variable, that leads me to ask, is this lane being run in parallel?

If so, then: 1) two or more `multi_scans` running in parallel _should_ make this error occur, 2) you can call `multi_scan` with the `:parallel_testrun_count` option with all the tests to run the tests evenly on each Simulator, and 3) if you really need to run `multi_scan` in parallel in a way that #2 is not valid, then you should specify a different `:output_directory` for each parallel job.

**Comment by lyndsey-ferguson** [comment]
See [this example](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/blob/master/lib/fastlane/plugin/test_center/actions/multi_scan.rb#L495) of running batches of tests in parallel. You don't even have to specify which tests to test, `multi_scan` figures it out.

**Comment by lyndsey-ferguson** [comment]
I have duplicated this issue.

**Comment by dp-1a** [comment]
Hi, sorry for the late response. 

Even without parallel jobs this issue seems to occur. We have the configuration you have recommended on our Jenkins based CI infrastructure and there it is guaranteed that only one job is running at the same time. I guess there must be some kind of race condition but unfortunately we where not able to narrow the issue further down.

Best regards

**Comment by dp-1a** [comment]
One rather hacky "fix" we tried, which seems to work but of course does not really fix the issue, was replacing `File.rename` with `FileUtils.mv(src_test_bundle, dst_test_bundle, :force => true)` on line 307 in the `retrying_scan_helper` file.

**Comment by lyndsey-ferguson** [comment]
@dp-1a & @rahulsharma2905, I believe that I have fixed the issue.

Can you modify your `Pluginfile` per my instructions below, run `bundle install`, and then run your fastlane again (with the `--verbose` flag)?

Pluginfile:
```
gem 'fastlane-plugin-test_center', :git => "https://github.com/lyndsey-ferguson/fastlane-plugin-test_center.git", :branch => "issue-169-fix-test_result_report_collation"
```

If there are still problems, please let me know and attach the console output as a text file to this issue (makes it easier for me to review).  If it works, please let me know.

**Comment by rahulsharma2905** [comment]
@lyndsey-ferguson, It worked without any error.

**Comment by dp-1a** [comment]
Hey @lyndsey-ferguson, your fix seems to work. The issue did not occur anymore in the latest runs. Thank you very much for your efforts.

**Comment by lyndsey-ferguson** [comment]
# 👏

**Comment by lyndsey-ferguson** [comment]
Fixed in [v3.8.14](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/releases/tag/v3.8.14)
